### ISSUE TITLE
playground: error when using 'show' function

### ISSUE THREAD

**Issue by srghma** [input-output-hk/plutus#1432]
I'm getting compilation error

```
GHC Core to PLC plugin: Error: Unsupported feature: Int: use Integer instead Context: Compiling definition of: GHC.Show.show Context: Compiling definition of: Main.correctGuess Context: Compiling definition of: Main.validateGuess Context: Compiling expr at "main:Main:(37,25)-(37,67)"
```

when trying to use `show` method in function

```
correctGuess :: HashedString -> ClearString -> Bool
correctGuess (HashedString actual) (ClearString guess') =
    traceH (show "asdf") False
```

expected - no error

tested on 
https://alpha.plutus.iohkdev.io/
https://testnet.iohkdev.io/plutus/tools/playground/

![Screenshot](https://user-images.githubusercontent.com/7573215/64888679-3da0a780-d674-11e9-8d5c-c5eabac4fa5c.png)

____

Full script

```

-- A game with two players. Player 1 thinks of a secret word
-- and uses its hash, and the game validator script, to lock
-- some funds (the prize) in a pay-to-script transaction output.
-- Player 2 guesses the word by attempting to spend the transaction
-- output. If the guess is correct, the validator script releases the funds.
-- If it isn't, the funds stay locked.
import qualified Language.PlutusTx          as PlutusTx
import qualified Language.PlutusTx
import           Language.PlutusTx.Prelude
import           Ledger                    
import           Ledger.Value               
import           Playground.Contract
import           Wallet                     
import qualified Data.ByteString.Lazy.Char8 as C

data HashedString = HashedString ByteString

PlutusTx.makeLift ''HashedString

data ClearString = ClearString ByteString

PlutusTx.makeLift ''ClearString

correctGuess :: HashedString -> ClearString -> Bool
correctGuess (HashedString actual) (ClearString guess') =
    traceH (show "asdf") False


validateGuess :: HashedString -> ClearString -> PendingTx -> Bool
validateGuess dataScript redeemerScript _ =
    correctGuess dataScript redeemerScript

-- | The validator script of the game.
gameValidator :: ValidatorScript
gameValidator =
    ValidatorScript ($$(Ledger.compileScript [|| validateGuess ||]))

-- create a data script for the guessing game by hashing the string
-- and lifting the hash to its on-chain representation
gameDataScript :: String -> DataScript
gameDataScript =
    DataScript . Ledger.lifted . HashedString . plcSHA2_256 . C.pack

-- create a redeemer script for the guessing game by lifting the
-- string to its on-chain representation
gameRedeemerScript :: String -> RedeemerScript
gameRedeemerScript =
    RedeemerScript . Ledger.lifted . ClearString . C.pack

-- | The address of the game (the hash of its validator script)
gameAddress :: Address
gameAddress = Ledger.scriptAddress gameValidator

-- | The "lock" contract endpoint. See note [Contract endpoints]
lock :: MonadWallet m => String -> Value -> m ()
lock word vl =
    -- 'payToScript_' is a function of the wallet API. It takes a script
    -- address, a currency value and a data script, and submits a transaction
    -- that pays the value to the address, using the data script.
    --
    -- The underscore at the end of the name indicates that 'payToScript_'
    -- discards its result. If you want to hold on to the transaction you can
    -- use 'payToScript'.
    payToScript_ defaultSlotRange gameAddress vl (gameDataScript word)

-- | The "guess" contract endpoint. See note [Contract endpoints]
guess :: (WalletAPI m, WalletDiagnostics m) => String -> m ()
guess word = do
    let redeemer = gameRedeemerScript word
    -- 'collectFromScript' is a function of the wallet API. It consumes the
    -- unspent transaction outputs at a script address and pays them to a
    -- public key address owned by this wallet. It takes the validator script
    -- and the redeemer scripts as arguments.
    --
    -- Note that before we can use 'collectFromScript', we need to tell the
    -- wallet to start watching the address for transaction outputs (because
    -- the wallet does not keep track of the UTXO set of the entire chain).
    collectFromScript defaultSlotRange gameValidator redeemer

-- | The "startGame" contract endpoint, telling the wallet to start watching
--   the address of the game script. See note [Contract endpoints]
startGame :: MonadWallet m => m ()
startGame =
    -- 'startWatching' is a function of the wallet API. It instructs the wallet
    -- to keep track of all outputs at the address. Player 2 needs to call
    -- 'startGame' before Player 1 uses the 'lock' endpoint, to ensure that
    -- Player 2's wallet is aware of the game address.
    startWatching gameAddress

$(mkFunctions ['lock, 'guess, 'startGame])

{- Note [Contract endpoints]

A contract endpoint is a function that uses the wallet API to interact with the
blockchain. We can look at contract endpoints from two different points of view.

1. Contract users

Contract endpoints are the visible interface of the contract. They provide a
UI (HTML form) for entering the parameters of the actions we may take as part
of the contract.

2. Contract authors

As contract authors we define endpoints as functions that return a value of
type 'MockWallet ()'. This type indicates that the function uses the wallet API
to produce and spend transaction outputs on the blockchain.

Endpoints can have any number of parameters: 'lock' has two
parameters, 'guess' has one and 'startGame' has none. For each endpoint we
include a call to 'mkFunction' at the end of the contract definition. This
causes the Haskell compiler to generate a schema for the endpoint. The Plutus
Playground then uses this schema to present an HTML form to the user where the
parameters can be entered.

-}
```

**Comment by srghma** [comment]
```
validateGuess :: HashedString -> ClearString -> PendingTx -> Bool
validateGuess dataScript redeemerScript p =
    traceH (show $ pendingTxHash p) (correctGuess dataScript redeemerScript)
```

doesnt work either

```
GHC Core to PLC plugin: Error: Unsupported feature: Kind: GHC.Types.RuntimeRep Context: Compiling definition of: GHC.Base.$ Context: Compiling definition of: Main.validateGuess Context: Compiling expr at "main:Main:(37,25)-(37,67)"
```

thought it has show instance

https://github.com/input-output-hk/plutus/blob/0e1576b598c111a409c042ed5f43f7cbea452164/plutus-wallet-api/ledger/Ledger/Validation.hs#L225-L229

**Comment by michaelpj** [comment]
The first error is strange. Are you sure you weren't doing `trace (show 1) ...`? Then the 1 might get defaulted to `Int` rather than `Integer`, which as the error message says, isn't supported.

The second error is a problem with `$`. The `$` exported by the Haskell prelude is ... very polymorphic, in ways that are hard to deal with. I've been meaning to export a simpler version from the Plutus Tx Prelude, I'll do that and that should make this okay.

**Comment by srghma** [comment]
@michaelpj 

> Are you sure you weren't doing trace (show 1) ...?

yes

here is the code

```

-- A game with two players. Player 1 thinks of a secret word
-- and uses its hash, and the game validator script, to lock
-- some funds (the prize) in a pay-to-script transaction output.
-- Player 2 guesses the word by attempting to spend the transaction
-- output. If the guess is correct, the validator script releases the funds.
-- If it isn't, the funds stay locked.

import qualified Language.PlutusTx          as PlutusTx
import           Language.PlutusTx.Prelude
import           Ledger                     (Address, DataScript (DataScript), PendingTx,
                                             RedeemerScript (RedeemerScript), ValidatorScript (ValidatorScript),
                                             compileScript, lifted, plcSHA2_256, scriptAddress)
import           Ledger.Value               (Value)
import           Playground.Contract
import           Wallet                     (MonadWallet, WalletAPI, WalletDiagnostics, collectFromScript,
                                             defaultSlotRange, payToScript_, startWatching)

import qualified Data.ByteString.Lazy.Char8 as C

data HashedString = HashedString ByteString

PlutusTx.makeLift ''HashedString

data ClearString = ClearString ByteString

PlutusTx.makeLift ''ClearString

correctGuess :: HashedString -> ClearString -> Bool
correctGuess (HashedString actual) (ClearString guess') =
    traceH (show "asdf"::String) False

validateGuess :: HashedString -> ClearString -> PendingTx -> Bool
validateGuess dataScript redeemerScript _ =
    correctGuess dataScript redeemerScript

-- | The validator script of the game.
gameValidator :: ValidatorScript
gameValidator =
    ValidatorScript ($$(Ledger.compileScript [|| validateGuess ||]))

-- create a data script for the guessing game by hashing the string
-- and lifting the hash to its on-chain representation
gameDataScript :: String -> DataScript
gameDataScript =
    DataScript . Ledger.lifted . HashedString . plcSHA2_256 . C.pack

-- create a redeemer script for the guessing game by lifting the
-- string to its on-chain representation
gameRedeemerScript :: String -> RedeemerScript
gameRedeemerScript =
    RedeemerScript . Ledger.lifted . ClearString . C.pack

-- | The address of the game (the hash of its validator script)
gameAddress :: Address
gameAddress = Ledger.scriptAddress gameValidator

-- | The "lock" contract endpoint. See note [Contract endpoints]
lock :: MonadWallet m => String -> Value -> m ()
lock word vl =
    -- 'payToScript_' is a function of the wallet API. It takes a script
    -- address, a currency value and a data script, and submits a transaction
    -- that pays the value to the address, using the data script.
    --
    -- The underscore at the end of the name indicates that 'payToScript_'
    -- discards its result. If you want to hold on to the transaction you can
    -- use 'payToScript'.
    payToScript_ defaultSlotRange gameAddress vl (gameDataScript word)

-- | The "guess" contract endpoint. See note [Contract endpoints]
guess :: (WalletAPI m, WalletDiagnostics m) => String -> m ()
guess word = do
    let redeemer = gameRedeemerScript word
    -- 'collectFromScript' is a function of the wallet API. It consumes the
    -- unspent transaction outputs at a script address and pays them to a
    -- public key address owned by this wallet. It takes the validator script
    -- and the redeemer scripts as arguments.
    --
    -- Note that before we can use 'collectFromScript', we need to tell the
    -- wallet to start watching the address for transaction outputs (because
    -- the wallet does not keep track of the UTXO set of the entire chain).
    collectFromScript defaultSlotRange gameValidator redeemer

-- | The "startGame" contract endpoint, telling the wallet to start watching
--   the address of the game script. See note [Contract endpoints]
startGame :: MonadWallet m => m ()
startGame =
    -- 'startWatching' is a function of the wallet API. It instructs the wallet
    -- to keep track of all outputs at the address. Player 2 needs to call
    -- 'startGame' before Player 1 uses the 'lock' endpoint, to ensure that
    -- Player 2's wallet is aware of the game address.
    startWatching gameAddress

$(mkFunctions 
    ['lock
    , 'guess
    , 'startGame
    ])

{- Note [Contract endpoints]

A contract endpoint is a function that uses the wallet API to interact with the
blockchain. We can look at contract endpoints from two different points of view.

1. Contract users

Contract endpoints are the visible interface of the contract. They provide a
UI (HTML form) for entering the parameters of the actions we may take as part
of the contract.

2. Contract authors

As contract authors we define endpoints as functions that return a value of
type 'MockWallet ()'. This type indicates that the function uses the wallet API
to produce and spend transaction outputs on the blockchain.

Endpoints can have any number of parameters: 'lock' has two
parameters, 'guess' has one and 'startGame' has none. For each endpoint we
include a call to 'mkFunction' at the end of the contract definition. This
causes the Haskell compiler to generate a schema for the endpoint. The Plutus
Playground then uses this schema to present an HTML form to the user where the
parameters can be entered.

-}
```

the error message is 

```
GHC Core to PLC plugin: Error: Unsupported feature: Int: use Integer instead Context: Compiling definition of: GHC.Show.show Context: Compiling definition of: Main.correctGuess Context: Compiling definition of: Main.validateGuess Context: Compiling expr at "main:Main:(40,25)-(40,67)"
```

tested here https://alpha.plutus.iohkdev.io/

**Comment by srghma** [comment]
> The second error is a problem with $. The $ exported by the Haskell prelude is ... very polymorphic, in ways that are hard to deal with. I've been meaning to export a simpler version from the Plutus Tx Prelude, I'll do that and that should make this okay.

here is the other code, without `$`

```

-- A game with two players. Player 1 thinks of a secret word
-- and uses its hash, and the game validator script, to lock
-- some funds (the prize) in a pay-to-script transaction output.
-- Player 2 guesses the word by attempting to spend the transaction
-- output. If the guess is correct, the validator script releases the funds.
-- If it isn't, the funds stay locked.

import qualified Language.PlutusTx          as PlutusTx
import           Language.PlutusTx.Prelude
import           Ledger                    
import           Ledger.Value               (Value)
import           Playground.Contract
import           Wallet                    

import qualified Data.ByteString.Lazy.Char8 as C

data HashedString = HashedString ByteString

PlutusTx.makeLift ''HashedString

data ClearString = ClearString ByteString

PlutusTx.makeLift ''ClearString

correctGuess :: HashedString -> ClearString -> Bool
correctGuess (HashedString actual) (ClearString guess') =
    False

validateGuess :: HashedString -> ClearString -> PendingTx -> Bool
validateGuess dataScript redeemerScript p =
    traceH (show (pendingTxHash p)) (correctGuess dataScript redeemerScript)

-- | The validator script of the game.
gameValidator :: ValidatorScript
gameValidator =
    ValidatorScript ($$(Ledger.compileScript [|| validateGuess ||]))

-- create a data script for the guessing game by hashing the string
-- and lifting the hash to its on-chain representation
gameDataScript :: String -> DataScript
gameDataScript =
    DataScript . Ledger.lifted . HashedString . plcSHA2_256 . C.pack

-- create a redeemer script for the guessing game by lifting the
-- string to its on-chain representation
gameRedeemerScript :: String -> RedeemerScript
gameRedeemerScript =
    RedeemerScript . Ledger.lifted . ClearString . C.pack

-- | The address of the game (the hash of its validator script)
gameAddress :: Address
gameAddress = Ledger.scriptAddress gameValidator

-- | The "lock" contract endpoint. See note [Contract endpoints]
lock :: MonadWallet m => String -> Value -> m ()
lock word vl =
    -- 'payToScript_' is a function of the wallet API. It takes a script
    -- address, a currency value and a data script, and submits a transaction
    -- that pays the value to the address, using the data script.
    --
    -- The underscore at the end of the name indicates that 'payToScript_'
    -- discards its result. If you want to hold on to the transaction you can
    -- use 'payToScript'.
    payToScript_ defaultSlotRange gameAddress vl (gameDataScript word)

-- | The "guess" contract endpoint. See note [Contract endpoints]
guess :: (WalletAPI m, WalletDiagnostics m) => String -> m ()
guess word = do
    let redeemer = gameRedeemerScript word
    -- 'collectFromScript' is a function of the wallet API. It consumes the
    -- unspent transaction outputs at a script address and pays them to a
    -- public key address owned by this wallet. It takes the validator script
    -- and the redeemer scripts as arguments.
    --
    -- Note that before we can use 'collectFromScript', we need to tell the
    -- wallet to start watching the address for transaction outputs (because
    -- the wallet does not keep track of the UTXO set of the entire chain).
    collectFromScript defaultSlotRange gameValidator redeemer

-- | The "startGame" contract endpoint, telling the wallet to start watching
--   the address of the game script. See note [Contract endpoints]
startGame :: MonadWallet m => m ()
startGame =
    -- 'startWatching' is a function of the wallet API. It instructs the wallet
    -- to keep track of all outputs at the address. Player 2 needs to call
    -- 'startGame' before Player 1 uses the 'lock' endpoint, to ensure that
    -- Player 2's wallet is aware of the game address.
    startWatching gameAddress

$(mkFunctions 
    ['lock
    , 'guess
    , 'startGame
    ])

{- Note [Contract endpoints]

A contract endpoint is a function that uses the wallet API to interact with the
blockchain. We can look at contract endpoints from two different points of view.

1. Contract users

Contract endpoints are the visible interface of the contract. They provide a
UI (HTML form) for entering the parameters of the actions we may take as part
of the contract.

2. Contract authors

As contract authors we define endpoints as functions that return a value of
type 'MockWallet ()'. This type indicates that the function uses the wallet API
to produce and spend transaction outputs on the blockchain.

Endpoints can have any number of parameters: 'lock' has two
parameters, 'guess' has one and 'startGame' has none. For each endpoint we
include a call to 'mkFunction' at the end of the contract definition. This
causes the Haskell compiler to generate a schema for the endpoint. The Plutus
Playground then uses this schema to present an HTML form to the user where the
parameters can be entered.

-}
```

the error message is 

```
GHC Core to PLC plugin: Error: Reference to a name which is not a local, a builtin, or an external INLINABLE function: Variable Data.ByteString.Lazy.toStrict_goLen0 No unfolding Context: Compiling definition of: LedgerBytes.$fShowLedgerBytes_$cshow Context: Compiling definition of: Main.validateGuess Context: Compiling expr at "main:Main:(37,25)-(37,67)"
```

**Comment by michaelpj** [comment]
Ah, the issue is you're using the Haskell Prelude `Show`, which doesn't work. 
- The first example uses `showsPrec` which uses `Int`.
- The second example uses some internal `ByteString` functoins.

Generally functions from other packages won't work since they're not typically `INLINABLE` (see the tutorial).

We don't have a great story for re-exported Haskell Prelude functions that don't work. I'll open another issue for that, maybe we should not export them...

**Comment by srghma** [comment]
Does it mean that in order to use `show` function I have to implement my own class `Show`?

**Comment by michaelpj** [comment]
Pretty much. String support on the Plutus side is *very* basic. It's mostly just limited to emitting literals at the moment, I think you'll rapidly run into the fact that there's no way to get a string representation of a integer or bytestring primitive.
